src/chain.h

```diff
@@ -7,15 +7,24 @@
 #define BITCOIN_CHAIN_H
 
 #include <arith_uint256.h>
-#include <consensus/params.h>
 #include <flatfile.h>
 #include <primitives/block.h>
+#include <serialize.h>
 #include <sync.h>
+#include <threadsafety.h>
 #include <uint256.h>
 #include <util/time.h>
 
+#include <cassert>
+#include <chrono>
+#include <cstdint>
+#include <string>
 #include <vector>
 
+namespace Consensus {
+struct Params;
```

Comment:
Why drop the `#include <consensus/params.h>` just to manually forward declare the struct? This seems like a backwards step. Seems better to either leave it as is, or move `GetBlockProofEquivalentTime` back to `pow.cpp` so that `chain.h` doesn't need `Consensus::Params` at all (partially reverting #7311 ; GBPET() is used in ConnectBlock() so it isn't really separate from consensus functionality for us).

Reply:
> Why drop the `#include <consensus/params.h>` just to manually forward declare the struct?

Although forward declarations have their own [pros and cons](https://google.github.io/styleguide/cppguide.html#Forward_Declarations), this code base seems to lean in their favour.

> This seems like a backwards step.

Mind elaborating?

---

As a question like that has been raised, should we document our preferences in https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md?

Reply:
> Mind elaborating?

If we've got a dependency from one module to another ("validation" depends on "consensus/params") then we should just include the header file; if we're avoiding a dependency (eg making "chain" just store chain data independent of whatever the consensus parameters might be), then we shouldn't be entangling them by referencing the "unrelated" module's definitions.

Some of those cases don't even avoid the dependency at all; eg validation.h includes chainparams.h which includes consensus/params.h anyway.

It makes the dependencies slightly harder to track: forward declarations are harder to search for than a plain #include; it also violates the DRY-principle, though not in a very big way.

The benefit is that not including the full header can reduce compilation time in the cases where you can avoid including a complicated header (eg, one that pulls in boost multiindex or that pulls in lots of other headers). That probably makes sense if you're avoiding pulling in txmempool.h (which validation.h does), but even then any effort's probably better spent in improving our modularisation...

Reply:
The changes in `src/chain.h` appeared non-required for this PR, and they have been [dropped](https://github.com/bitcoin/bitcoin/pull/26763#issuecomment-1370769788).

The discussion about forward declaration vs header including could be continued elsewhere, maybe in a dedicated issue, couldn't it?